### ISSUE TITLE
Memory growth and JS

### ISSUE THREAD

**Issue by kripken** [WebAssembly/WASI#82]
It looks like the current wasi libc implements sbrk using the clang builtin to grow, and there isn't a wasi API for growth. I think this may be a problem for a JS embedding (that is, running a wasi program with JS implementing the wasi APIs etc.), as any JS views on the buffer used in the wasm Memory will become invalid - they don't resize automatically, and must be manually recreated. In particular I think the current Web polyfill for wasi probably doesn't fully work with memory growth.

There is no event callback for when a Memory grows, but even if there were, it wouldn't be enough, just like [with pthreads](https://github.com/WebAssembly/design/issues/1271) - the event would happen on a later JS event loop iteration, and not when we need it.

One possible solution here would be to add an API to wasi that either does the growth (`__wasi_grow_memory`?), or that notifies the runtime about the growth (`__wasi_notify_memory_growth`?).

**Comment by kripken** [comment]
@alexcrichton I'm skeptical of the JS side having to check because it means every single location that calls into wasm needs extra work, and if you forget one you get breakage later. That would make it much more clumsy to use wasi binaries from JS...

@devsnek I'm also skeptical of a synchronous callback - is that common on the Web? Offhand the only example I can think of is sorting, and it's not even a real callback there, since the entire thing happens immediately...

**Comment by guybedford** [comment]
@kripken script load callbacks on the web are synchronous.

**Comment by devsnek** [comment]
@kripken splitting your question into two parts:

EventTarget (web) and EventEmitter (node) dispatch events synchronously, but calls dispatch those events are generally queued via the event loop.

the wasm spec doesn't assume anything about the environment besides it being javascript, so we can't even use EventTarget or EventEmitter anyway, and so i think any precedence using them might bring is moot.

**Comment by sbc100** [comment]
@kripken are you worried about the cost of the `if (mem.buffer)` at the start of each JS syscall?  Or are you worried about all the other JS functions that emscripten users might want to expose to wasm?

If its the latter then presumably we can wrap all JS functions that we export to wasm a helper that does this?   It would be useful to have that ability anyway for other things such as tracing that wasm boundary.

**Comment by kripken** [comment]
Thanks @devsnek! It does sound like that may be a good option then. I opened https://github.com/WebAssembly/design/issues/1296 , please let me know if I wrote this up ok.

@sbc100 yeah, but not just syscalls - any time you enter JS, if wasm might have run, you need to do that, and any time right after you call into wasm and return from there.

So if someone loads a wasi wasm file in JS, and interleaves calling exports from there with JS looking at the memory, bugs can easily happen. Like imagine box2d.wasm,

```
jsGlue.addObject(); // JS looking at views
box2D.calculatePhysics(); // wasm runs
jsGlue.readWorldState(); // JS looking at views
```
For that to be correct, we'd need to check if memory grew right after that wasm call. In other words, users need to be very careful...

**Comment by sbc100** [comment]
@kripken could we not do that automatically since we control the list of functions we give to wasm that the functions we get back.

It would mean that the functions we expose to user JS are not that actual wasm functions but wrappers.  And the JS function we give to wasm would also be wrappers.  

This is obviously overhead, I'm just not sure how much given that the crossing from JS to wasm presumably already has a fair amount.

**Comment by kripken** [comment]
@sbc100 Yes, some automatic instrumentation is possible here. But it means that every user that uses wasi in JS must remember to do that, and do it properly - this isn't just for emscripten users!

**Comment by sbc100** [comment]
Emscripten is a little different in that its users tend to expose arbitrary functions to their wasm programs.  In the WASI world the idea is that the program will only be exposed to the WASI syscall APIs, not arbitrary host functions.

I'm not saying the current situation is ideal.  Its certainly on overhead for JS embedders that they have to handle this, but hopefully its something that can be done at embedder level not something that individual developer or app will need to care about.

**Comment by kripken** [comment]
> In the WASI world the idea is that the program will only be exposed to the WASI syscall APIs, not arbitrary host functions.

Will wasi programs not want to provide exports that can be called from JS?

**Comment by sunfishcode** [comment]
At the moment, all WASI programs are Commands, which means they just have a `_start` function that needs to get called. We're working on establishing Reactors and Libraries as two additional forms of WASI programs, which would support calls from the outside, but at the moment those concepts are still being developed.

**Comment by devsnek** [comment]
Another solution might be interface types. If wasi apis passed live memory views instead of pointers, this wouldn't be a problem anymore.

**Comment by skoppe** [comment]
> There is no event callback for when a Memory grows

I have had the same problem in [skoppe/spasm](https://github.com/skoppe/spasm) and - like everyone else - ended up using a check at the beginning of each JS function.

In my case I was statically linking imgui (compiled to wasm with wasi-libc) with a webgl backend written in Dlang. Both the D program and wasi-libc were issuing calls to the memory grow intrinsic, at which point I realised I couldn't intercept calls to the grow intrinsic in the D program, since it would miss those issued from wasi-libc.

So I ended up injecting memory checks in each js function. It didn't feel right when I wrote it, and it still doesn't.

While I understand that a reallocation invalidates pointers, I am of the opinion that the wasm runtime ought to fix that pointer after a memory grow.

To check whether the typed js array is still valid I check the length property, which equals 0 when the memory is invalid. This suggests to me there is already some bounds checking and memory validation code whenever JS accesses the typed array. If that is the case, having the wasm runtime update the typed array when a memory grow happens should cost very little.
